File: python-algo/leetcode/problem_2707.py
"""2707. Extra Characters in a String
https://leetcode.com/problems/extra-characters-in-a-string/
"""
from typing import List
from functools import cache


class Solution:
    def minExtraChar(self, s: str, dictionary: List[str]) -> int:
        @cache
        def dfs(pre: str, i: int) -> int:
            if i == len(s):
                if not pre or pre in d:
                    return 0
                return len(pre)
            cur = pre + s[i]
            ret = dfs(cur, i + 1)
            if cur in d:
                ret = min(ret, dfs('', i + 1))
            else:
                ret = min(ret, dfs(s[i], i + 1) + len(pre), dfs('', i + 1) + len(cur))
            return ret

        d = set(dictionary)
        return dfs('', 0)

File: java-algo/src/main/java/com/leetcode/Problem2707.java
package com.leetcode;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

/**
 * 2707. Extra Characters in a String
 * https://leetcode.com/problems/extra-characters-in-a-string/
 */
public class Problem2707 {
    public int minExtraChar(String s, String[] dictionary) {
        int n = s.length();
        int[] dp = new int[n + 1];
        Arrays.fill(dp, n + 1);
        dp[0] = 0;
        Set<String> dicts = new HashSet<>(Arrays.asList(dictionary));
        for (int i = 1; i <= n; i++) {
            dp[i] = dp[i - 1] + 1;
            for (int j = i - 1; j >= 0; j--) {
                if (dicts.contains(s.substring(j, i))) {
                    dp[i] = Math.min(dp[i], dp[j]);
                }
            }
        }
        return dp[n];
    }
}
